tools/gfx/d3d11/render-d3d11.cpp

```diff
@@ -542,22 +423,1101 @@ class D3D11Device : public ImmediateRendererBase
+            SLANG_RETURN_ON_FAIL(device->createBufferResource(
+                IResource::Usage::ConstantBuffer, bufferDesc, nullptr, bufferResourcePtr.writeRef()));
+            m_ordinaryDataBuffer = static_cast<BufferResourceImpl*>(bufferResourcePtr.get());
+
+            // Once the buffer is allocated, we can use `_writeOrdinaryData` to fill it in.
+            //
+            // Note that `_writeOrdinaryData` is potentially recursive in the case
+            // where this object contains interface/existential-type fields, so we
+            // don't need or want to inline it into this call site.
+            //
+            SLANG_RETURN_ON_FAIL(_writeOrdinaryData(device, m_ordinaryDataBuffer, 0, specializedOrdinaryDataSize, specializedLayout));
+
+            return SLANG_OK;
+        }
+
+        /// Bind the buffer for ordinary/uniform data, if needed
+        Result _bindOrdinaryDataBufferIfNeeded(
+            D3D11Device* device,
+            RootBindingState* bindingState)
+        {
+            // We start by ensuring that the buffer is created, if it is needed.
+            //
+            SLANG_RETURN_ON_FAIL(_ensureOrdinaryDataBufferCreatedIfNeeded(device));
+
+            // If we did indeed need/create a buffer, then we must bind it
+            // into root binding state.
+            //
+            if (m_ordinaryDataBuffer)
+            {
+                bindingState->constantBuffers.add(m_ordinaryDataBuffer->m_buffer);
+            }
+
+            return SLANG_OK;
+        }
+    public:
+        virtual Result bindObject(D3D11Device* device, RootBindingState* bindingState)
+        {
+            ShaderObjectLayoutImpl* layout = getLayout();
+
+            Index baseRangeIndex = 0;
+            SLANG_RETURN_ON_FAIL(_bindOrdinaryDataBufferIfNeeded(device, bindingState));
+
+            for (auto sampler : m_samplers)
+                bindingState->samplerBindings.add(sampler ? sampler->m_sampler.get() : nullptr);
+            
+            for (auto srv : m_srvs)
+                bindingState->srvBindings.add(srv ? srv->m_srv : nullptr);
+
+            for (auto uav : m_uavs)
+                bindingState->uavBindings.add(uav ? uav->m_uav : nullptr);
+
+            for (auto subObject : m_objects)
+                subObject->bindObject(device, bindingState);
+            
+            return SLANG_OK;
+        }
+
+        /// Any "ordinary" / uniform data for this object
+        List<char> m_ordinaryData;
+
+        List<RefPtr<ShaderResourceViewImpl>> m_srvs;
+
+        List<RefPtr<SamplerStateImpl>> m_samplers;
+
+        List<RefPtr<UnorderedAccessViewImpl>> m_uavs;
+
+        List<RefPtr<ShaderObjectImpl>> m_objects;
+
+        /// A constant buffer used to stored ordinary data for this object
+        /// and existential-type sub-objects.
+        ///
+        /// Created on demand with `_createOrdinaryDataBufferIfNeeded()`
+        RefPtr<BufferResourceImpl> m_ordinaryDataBuffer;
```

Comment:
Again, this is awesome to see and exactly what we'd want/expect in this case.

Maybe the only detailed difference we could look for is to have a flat array next to `m_srvs` that holds just a bunch of `ID3D11ShaderResourceView*` so that it can be passed directly to API calls with no extra translation.